### ISSUE TITLE
Question - Whats the best way to access a method inside a Child Step component

### ISSUE THREAD

**Issue by newbreedofgeek** [Srdjan/react-multistep#17]
Hello,

This is more of a question than a issue.

I'm using your example code here:
https://github.com/Srdjan/react-multistep/tree/master/example/app

But as the user clicks on the "Next" button I need to first call a method inside the child (step) component before rendering it.

For example, I have a isValidated() method inside each of my child (step) components. When you click on the next button I want to first check if the current step has validated and then move on.

e.g.:
```
next() {
    if (this.props.steps[this.state.compState].component.isValidated()) {
       this.setNavState(this.state.compState + 1)
    }
}

```
I want to do this without using a store (flux / redux events) as this has more to do with component state rather than data state.

Once way is to somehow assign a "ref" to the child step component in the render() of MultiStep.

e.g.
```
render() {
    return (
        ...
        { this.props.steps[this.state.compState].component}.refs = 'activeComponent' }
        ...
    )
}
```

and then in next() access the activeComponent's isValidated() method.

```
next() {
    if (this.refs.activeComponent.isValidated()) {
       this.setNavState(this.state.compState + 1)
    }
}
```

But unfortunately I am not sure if we can add a refs to the step property as the child step component is passed in the array.

How would you do this? any advise / help?

**Comment by Srdjan** [comment]
so, my head is out of React world for a while now...
but, here are my thoughts (@jaimeagudo and @kaadash may have better alternatives):

1) we could add 'validate' function, as in your example as a second export from the component:
'use strict'
import React, { Component, PropTypes } from 'react'

const store = { checked: false }
const validate = function() {...} 
const StepFour = React.createClass ({
...
}
export { StepFour, validate }

than:
import { StepOne, validate as validateStepOne } from './stepone'
import { StepTwo, validate as validateStepTwo } from './steptwo'
import { StepThree, validate as validateStepThree } from './stepthree'
import { StepFour, validate as validateStepFour } from './stepfour'

const steps = 
    [
      {name: 'StepOne', component: <StepOne/>, isValidated: validateStepOne},
      {name: 'StepTwo', component: <StepTwo/>, isValidated: validateStepTwo},
      {name: 'StepThree', component: <StepThree/>, isValidated: validateStepThree},
      {name: 'StepFour', component: <StepFour/>, isValidated: validateStepFour}
    ]
and invoke it the way you proposed...

2) Consider validation at the field level, so that the user is immediately notified. For that you would  apply validation within the component internally - and in addition to flag the field also raise an event that would signal parent to disable navigation since the current field holds invalid state 

Hope this helps

**Comment by newbreedofgeek** [comment]
Thanks for the advise. 

I did work out an interim solution by cloning the step component and extending it with a 'ref' 

```
render() {
    let compToRender = React.cloneElement(this.props.steps[this.state.compState].component, {
        ref: 'activeComponent'
    });
...
    { compToRender }
...
}
```
Which then gave me access to the public methods of the step component, which has a isValidated() method.

```
if (this.refs.activeComponent.isValidated()) {
     this.setNavState(this.state.compState + 1);
}
```


Thanks again,

**Comment by newbreedofgeek** [comment]
Happy with the answers...

**Comment by Srdjan** [comment]
great - if you end up having something that can be shared - happy to take PR :)

**Comment by jaimeagudo** [comment]
I'd disable the next-previous buttons on the component itself and would render them externally as children of the component so you can fine control the behaviour and drive the change using the `activeStep` prop, `refs` usage is discouraged @newbreedofgeek 

https://github.com/jaimeagudo/react-multistep/blob/master/src/index.js#L114

**Comment by newbreedofgeek** [comment]
@Srdjan thanks. I did fork your codebase and created a modified version that I'm actively maintaining now as i'm using it heavily in a project (it's got validation invocation as well as more options based control). 
https://github.com/newbreedofgeek/react-stepzilla

@jaimeagudo thanks for your recommendation and agree that `refs` should be avoided. My intial design had the next-prev buttons in the Children but in a case where you have many Child step components that design is not maintainable and not very DRY. So I used the original design and 'refed'.. not very happy with it but it'll do until I find a better solution.
